### PR TITLE
Add support for returning alternative tokens

### LINKED ISSUE #298: Include logprobs for alternative, less probable tokens in the generation response
### Feature request

For some applications it is necessary to access the full probability distribution of the tokens generated by the model's decoder head. This could be implemented by including a (size-limited) list of alternatives with every generated token in the `details` field of a response. This list would then include the `k` most probable tokens for each generation step, including the logprob and decoded text of the token.

### Motivation

Having access to the full probability distribution of the LLM decoder head allows for a more precise analysis of the model's response. For example, by comparing the logprob of the chosen token with the alternatives, we can deduce how likely the model was to pick a different option for that next token.

### Your contribution

For models based on `FlashCausalLM`, the necessary changes are implemented in #297. Here, the client parameter `return_k_alternatives` can be used to set the size of the list of alternatives that is returned with each generated token. For models based on other classes, setting this option currently has no effect.

### PR DESCRIPTION
**Pull request by JTS22**
# Add support for returning alternative tokens

Fixes #298
So far, alternative tokens are only supported on `FlashCausalLM` based models.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Was this discussed/approved via a Github issue or the discord / slack channel? Please add a link
      to it if that's the case: #298 
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: server/lorax_server/models/flash_causal_lm.py [hunk 1]
```diff
@@ -1014,6 +1023,33 @@ def generate_token(
...
+                for alternative_token_id in batch_alternative_token_ids:
+                    all_input_ids.append(alternative_token_id)
+                    alternative_token_text, _, _ = self.decode_token(
+                        all_input_ids,
+                        prefix_offset,
+                        read_offset,
+                    )
+                    alternative_token_texts.append(alternative_token_text)
+                    all_input_ids.pop()
+                alternative_tokens = AlternativeTokens(
```
Comment (tgaddair) [inline server/lorax_server/models/flash_causal_lm.py:1045]: It looks like we're overriding this at every iteration of the loop. Should `alternative_tokens` instead be a list of `AlternativeTokens`, one for each request in the batch?

**Review by tgaddair — COMMENTED** [review]
Very clean and comprehensive PR, just had a question about how we handle multiple requests in the batch.

File: server/lorax_server/models/flash_causal_lm.py (hunk 1 above)
Comment (JTS22) [reply server/lorax_server/models/flash_causal_lm.py:1045]: We are adding the `alternative_tokens` to the `Generation` that is created for each request, so I don't see a problem with overriding it. Basically I just copied from what is done with the `PrefillTokens`. The variables that go into `AlternativeTokens` are kind of badly named though, so I added another commit to fix that :)

**Review by tgaddair — APPROVED** [review]
LGTM! Thanks for this contribution :)

**Pull request merged**